src/win32/win32_socket.c

```diff
@@ -0,0 +1,316 @@
+
+    return(rc);
+}
+
+ssize_t win32_read_socket(SOCKET fd, void *buf, int n) {
+    int rc = recv(fd, buf, n, 0);
+
+    if (rc == SOCKET_ERROR) { set_socket_errno(WSAGetLastError()); }
+
+    return(rc);
+}
+
+
+char * win32_strtok_r(char *s, const char *delim, char **lasts) {
+    register char *spanp;
+    register int c, sc;
+    char *tok;
+
+
+    if (s == NULL && (s = *lasts) == NULL) { return (NULL); }
+
+    // skip (span) leading delimiters (s += strspn(s, delim), sort of).
+cont:
+    c = *s++;
+
+    for (spanp = (char *)delim; (sc = *spanp++) != 0;)
+        if (c == sc) { goto cont; }
+
+
+    if (c == 0) { // no non-delimiter characters
+        *lasts = NULL;
+        return (NULL);
+    }
+
+    tok = s - 1;
+
+    /* Scan token (scan for delimiters: s += strcspn(s, delim), sort of).
+     * Note that delim must have one NUL; we stop if we see that, too.
+     */
+    for ( ; ;) {
+        c = *s++;
+        spanp = (char *)delim;
+
+        do
+            if ((sc = *spanp++) == c) {
+                if (c == 0) {
+                    s = NULL;
+                } else {
+                    s[-1] = 0;
+                }
+
+                *lasts = s;
+                return(tok);
+            }
+
+        while (sc != 0);
+    }
+
+    // NOT REACHED
+}
+
+char *win32_strsep (char **stringp, const char *delim) {
+    register char *s;
+    register const char *spanp;
+    register int c, sc;
+    char *tok;
+
+    if ((s = *stringp) == NULL) {
+        return(NULL);
+    }
+
+    for (tok = s; ;) {
+        c = *s++;
+        spanp = delim;
+
+        do
+            if ((sc = *spanp++) == c) {
+                if (c == 0) {
+                    s = NULL;
+                } else {
+                    s[-1] = 0;
+                }
+
+                *stringp = s;
+                return(tok);
+            }
+
+        while (sc != 0);
+    }
+
+    // NOT REACHED
+}
+
+#ifndef STLINK_HAVE_UNISTD_H
+int usleep(unsigned int waitTime) {
+    if (waitTime >= 1000) {
+        /* Don't do long busy-waits.
+         * However much it seems like the QPC code would be more accurate,
+         * you can and probably will lose your time slice at any point during the wait,
+         * so we might as well voluntarily give up the CPU with a WaitForSingleObject.
+         */
+        HANDLE timer;
+        LARGE_INTEGER dueTime;
+        dueTime.QuadPart = -10 * (LONGLONG)waitTime;
+        timer = CreateWaitableTimer(NULL, TRUE, NULL);
+        SetWaitableTimer(timer, &dueTime, 0, NULL, NULL, 0);
+        WaitForSingleObject(timer, INFINITE);
+        CloseHandle(timer);
+
+        return(0);
+    }
+
+    LARGE_INTEGER perf_cnt, start, now;
+
+    QueryPerformanceFrequency(&perf_cnt);
+    QueryPerformanceCounter(&start);
+
+    do
+        QueryPerformanceCounter((LARGE_INTEGER*)&now);
+    while ((now.QuadPart - start.QuadPart) / (float)perf_cnt.QuadPart * 1000 * 1000 < waitTime);
```

Comment:
Think there is also need brackets for "do while" loop.

Reply:
Not only there, within this file. Will be fixed in the next commit.